### ISSUE TITLE
Clean up libdvd* usage: Upstream fixes and stdio override

### ISSUE THREAD

**Issue by pkerling** [xbmc/xbmc#15899]
## Roadmap or todo item
**Detailed Description**

Kodi has private copies of libdvdread, libdvdcss, and libdvdnav. They contain many patches that were never upstreamed. It is unclear how many of them are still needed.

These libraries are loaded from `.dll`. or `.so` files and have their stdio functions hooked in order to use the Kodi vfs.

We should rectify this at some point.

### Context

Keeping a patched version of the libdvd* libraries in our source tree is not the cleanest and most maintainable solution. stdio hooking is very hacky.

### Possible Implementation

* Add hooks to libdvd* to allow using vfs
* Use these hooks and drop the stdio override code in Kodi
* Go through all patches to libdvd* libs and decide whether to try to upstream or drop them

### Additional context, screenshots or links
Here are some relevant links or screenshots

https://code.videolan.org/videolan/libdvdread/issues/3

*Note: Team Kodi is considering this item, but we are not making any promises if and when it will be included.*

**Comment by davilla** [comment]
https://github.com/MrMC/mrmc/tree/master/lib/libdvd

see dvd_register_dir and dvd_register_file.

libdvd was moved into core and modernized because upstream for it is dead.

**Comment by pkerling** [comment]
Many thanks for the libdvd link, I didn't know about that. Am I interpreting this correctly that you (or whoever) consolidated the libdvd* libraries into one and have working IO hooks for directories already?

> libdvd was moved into core and modernized because upstream for it is dead.

I've talked to JB personally last year or so and he said that they are willing to merge anything reasonable (i.e. fixes a real problem and code is clean). It is possible to open merge requests on GitLab now. The problem with the last upstreaming attempt done by us was that it was just one huge PR with all our assorted changes and no rationale, IIRC, so it got rejected on that grounds.

**Comment by davilla** [comment]
Yes, I removed their 'streaming' addition they added last year. It was ok but very limited in ability.
The problem with merging clean is that it is a twisted mess in there :)
 
Added in VFS handling based on how libbluray does it, the 1st implementation of VFS in libbluray.

Then removed the need for libdvdcss. Most users already strip any encryption and libdvdcss was messy in trying to adapt to using the VFS.

Then combined libdvdnav and libdvdread and fixed up the mess of includes. 
In other words, dumped all into a blender and skimmed off the cream :)

My perspective, libdvdnav, libdvdread and libdvdcss has served their functions well over the years and it was time to refactor the cruft away. It has been two years since this refactor and we never get reports of dvd issues for ISO's or VIDEO_TS handling.

It was moved into core to make it easier to dev/debug, it could come out now and go back to a lib. I think. It works fine now so I see no pressing need to diddle with it.

**Comment by pkerling** [comment]
Sounds great so far, but:

> Then removed the need for libdvdcss. Most users already strip any encryption and libdvdcss was messy in trying to adapt to using the VFS.

Does this mean that users will be unable to play physical (CSS-encrypted) DVDs?

**Comment by davilla** [comment]
Correct, they would have to rip/decrypt the dvd using some other App. This is the dominant use case for iOS/tvOS/Droid platforms and I suspect most others too.

**Comment by pkerling** [comment]
I'm not sure we want to go this way for PC-based platforms then

**Comment by davilla** [comment]
Sure, if you desire to retain libdvdcss functionality, then our changes will not provide a solution.

**Comment by pkerling** [comment]
Mh how difficult do you think it would be to (re)integrate CSS? (all things considered it might still be easier than to do the required changes to libdvd*, that's why I'm asking)

**Comment by davilla** [comment]
Non-trivial but could be done. The main issue is that each of the three libs assume that file i/o is local and the structures are just not present to link them back to a context VFS handler without the icky usage of globals.
libdvdcss was the worst offender which is why it was dropped.
